src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/UseNameofInPlaceOfString.cs

```diff
@@ -0,0 +1,141 @@
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftMaintainabilityAnalyzersResources.UseNameOfInPlaceOfStringMessage), MicrosoftMaintainabilityAnalyzersResources.ResourceManager, typeof(MicrosoftMaintainabilityAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftMaintainabilityAnalyzersResources.UseNameOfInPlaceOfStringDescription), MicrosoftMaintainabilityAnalyzersResources.ResourceManager, typeof(MicrosoftMaintainabilityAnalyzersResources));
+
+        internal static DiagnosticDescriptor RuleWithSuggestion = new DiagnosticDescriptor(RuleId,
+                                                                         s_localizableTitle,
+                                                                         s_localizableMessage,
+                                                                         DiagnosticCategory.Maintainability,
+                                                                         DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                         isEnabledByDefault: true,
+                                                                         description: "Use nameof",
+                                                                         // TODO: add MSDN url
+                                                                         helpLinkUri: "http://msdn.microsoft.com/library/ms182181.aspx",
+                                                                         customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleWithSuggestion);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationAction(AnalyzeArgument, OperationKind.Argument);
+        }
+
+        private void AnalyzeArgument(OperationAnalysisContext context)
+        {
+
+            var argument = (IArgumentOperation)context.Operation;
+
+            if (argument.Value.Kind != OperationKind.Literal)
+            {
+                return;
+            }
+            // TODO better way to get the string?
+            var stringText = argument.Value.ConstantValue.Value.ToString();
+            var properties = ImmutableDictionary<string, string>.Empty.Add("StringText", stringText);
+
+            // TODO when showing diagnostic on a named argument, should just squiggle the string and not the argument name
+            var matchingParameter = (IParameterSymbol)argument.Parameter;
```

Comment:
Can the matching parameter be null? For example, for error cases where you are passing an additional string argument, when there is no matching parameter? Do add a unit test to cover the case.

Reply:
@mavasani , when I pass three strings to ArgumentNullException, processing stops after this line: ` analysisContext.RegisterOperationAction(AnalyzeArgument, OperationKind.Argument);`  Is this because the method was not able to bind to an overload and therefore it is not of OperationKind.Argument?

Reply:
Also, I added the test case, and it is passing - no diagnostic is shown and no errors.